pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,443 @@
+			}); err != nil {
+				t.Fatal(err)
+			}
+
+			t.L().Printf("deleted all data using tombstones")
+
+			if err := checkTableHasNoLiveData(t, conn, "kv", "kv"); err != nil {
+				t.Fatal(err)
+			}
+
+			execSQLOrFail("alter database kv configure zone using gc.ttlseconds = $1", 120)
+
+			c.Run(ctx, c.Node(1), "./cockroach", "workload", "run", "kv", "--duration", "2m",
+				"--read-percent", "10")
+
+			t.L().Printf("run batch of workload on top of deleted data")
+
+			deadline := timeutil.Now().Add(5 * time.Minute)
+			gcRetryTime := timeutil.Now()
+			for {
+				if timeutil.Now().After(gcRetryTime) {
+					enqueueAllTableRangesForGC(ctx, t, c, m)
+					gcRetryTime = timeutil.Now().Add(2 * time.Minute)
+				}
+				err := assertTableHasNoGarbage(t, conn, "kv", "kv")
+				if err == nil {
+					break
+				}
+				if timeutil.Now().After(deadline) {
+					t.Fatalf("tombstones were not GCd: %s", err)
+				}
+				select {
+				case <-time.After(10 * time.Second):
+				case <-ctx.Done():
+					// Test is cancelled.
+					return nil
+				}
+			}
+
+			t.L().Printf("all tombstones were garbage collected")
+		}
+
+		return nil
+	})
+	m.Wait()
+}
+
+func splitKvIntoRanges(
+	t test.Test, conn *gosql.DB, ranges int, expiration time.Duration, tableName, databaseName string,
+) {
+	if ranges < 2 {
+		return
+	}
+	var values = make([]string, ranges-1)
+	for i := 0; i < ranges-1; i++ {
+		values[i] = fmt.Sprintf("(%d)", getSplitKey(i+1, ranges))
+	}
+	stmt := fmt.Sprintf(`alter table %s.%s split at values %s with expiration now() + INTERVAL '%d'`,
+		databaseName, tableName, strings.Join(values, ", "), int(expiration.Seconds()))
+	if _, err := conn.Exec(stmt); err != nil {
+		t.Fatalf("failed to split kv table into ranges: %s", err)
+	}
+}
+
+// Get split point splits uint64 key space into 'fragments' chunks and returns
+// 'index' chunk start.
+func getSplitPoint(index, fragments int) uint64 {
+	return math.MaxUint64 / uint64(fragments) * uint64(index)
+}
+
+// Get split keys gives back int64 split points for keys used by kv workload.
+func getSplitKey(index, fragments int) int64 {
+	return math.MinInt64 + int64(getSplitPoint(index, fragments))
+}
+
+func assertTableHasNoGarbage(t test.Test, conn *gosql.DB, tableName, databaseName string) error {
+	totals, rangeCount, sts := collectTableMVCCStatsOrFatal(t, conn, tableName, databaseName)
+	t.L().Printf("assert range tombstones and garbage in ranges: %d stats: %s", rangeCount, totals.String())
+	if rangeCount == 0 {
+		return errors.New("failed to find any ranges for table")
+	}
+	if totals.RangeKeyCount > 0 || totals.RangeKeyBytes > 0 {
+		return errors.Errorf("table ranges contain range tombstones %s", totals.String())
+	}
+	if totals.GCBytesAge != 0 || totals.GCBytes() > 0 {
+		return errors.Errorf("table ranges contain garbage %s", totals.String())
+	}
+	if c := sts[roachpb.CheckConsistencyResponse_RANGE_INCONSISTENT.String()]; c > 0 {
```

Comment:
This isn't really the status we care about. It will only detect data inconsistencies, which practically never happen. We're specifically interested in incorrect stats, i.e. `RANGE_CONSISTENT_STATS_INCORRECT`. But we should error on any status other than `RANGE_CONSISTENT`.

Reply:
Good point, but I we also get `RANGE_INDETERMINATE` if consistency check times out internally which I saw quite a bit when running it.
Should we fail on `RANGE_CONSISTENT_STATS_INCORRECT` and `RANGE_INCONSISTENT` while ignore `RANGE_INDETERMINATE` and `RANGE_CONSISTENT_STATS_ESTIMATED`?

Reply:
Do you know why we get indeterminate? I thought we typically got that for unavailable replicas, are there any unavailable replicas/nodes here? When I added this in kvnemesis I never saw that happen.

I think we can error on stats estimates too. None of the operations here should have estimated stats, afaik.

Reply:
As discussed offline that we likely due to underreplication on startup amplified by agressive pre splits.
Test now expects consistency for all ranges at the end. There's no point for checking consistently multiple times as it is not likely that consistency will get fixed by itself it if becomes inconsistent at some point.